enhancements/network/ingress-fault-detection.md

```diff
@@ -0,0 +1,163 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement enriches the [Ingress Operator](https://github.com/openshift/cluster-ingress-opreator)
+by granting the operator the ability to automatically detect when ingress-related issues occur within a cluster.
+This enhancement also gives Cluster Admins the ability to pinpoint which specific part of the cluster Ingress "chain" is at fault when
+external traffic bound for application pods is unable to reach (or experiences difficulties en-route to) the correct application destination.
+
+## Motivation
+
+Cluster Ingress issues can greatly and rapidly impact the health of an OpenShift cluster. Cluster Administrators should be alerted to
+any potential ingress complications through the Ingress Operator as soon as they occur,
+as to help minimize unplanned application downtime. OpenShift cluster administrators should not first find out about ingress breakages
+when they inspect the status of a different critical cluster component, such as the console. A Cluster Administrator should also
+have tooling available to them to thoroughly diagnose and understand why routes on their cluster are not working.
+
+This enhancement also has benefits for internal OpenShift development.
+Often times when networking-related bugs are identified within a cluster, determining which networking sub-component
+(router, SDN, DNS, etc.) is responsible for the observed bug is difficult. Bug reports tend to spend lots of time bouncing between different teams
+as engineers are not sure which component(s) is legitimately at fault. Being able to quickly pinpoint Ingress related issues will expedite the bugfix process.
+
+### Goals
+
+1. Teach the [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) to automatically detect cluster Ingress issues
+and alert cluster administrators when relevant problems are discovered.
+2. Provide a secondary tool that can be triggered at will by a Cluster Administrator or an OpenShift support engineer that spawns an Ingress "debug"
+container that quickly identifies and triages potential networking issues within the cluster.
+
+### Non-Goals
+
+1. Teach the Ingress Operator to "self-heal" when cluster Ingress errors or issues are observed.
+2. Modify the [Ingress Controller API](https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go)
+to allow Cluster Administrators to disable the periodic Ingress Operator fault detector at will for a given Ingress Controller.
+
+## Proposal
+
+Once the Ingress Operator has successfully deployed an [Ingress Controller](https://github.com/openshift/cluster-ingress-operator/blob/master/README.md)
+within a cluster, the Ingress Operator will spawn a simple HTTP echo container, such as [Hello Openshift](https://hub.docker.com/r/openshift/hello-openshift/).
```

Comment:
Before moving in this direction, we probably want to get hello-openshift images for different architectures ([rhbz#1766287][1]).  Although I guess we'll be adding this image to the core images referenced from the release image, so it gets mirrored into disconnected environments, and in that case ART's current arch-specific images feeding arch-specific release images should cover the arch issue.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1766287

Reply:
@wking thanks for bringing this concern to my attention. I feel that `hello-openshift` should become a part of the release image if and when we go forward with this proposal, given that `hello-openshift` is only ~ 6 mb in size, and that it is useful as a testing application for OCP users.

Alternatively, the Ingress fault detection setup could use a base image already included within the release image, and create a simple socat echo server at runtime on top of the base image, as seen in the [router's CI](https://github.com/openshift/origin/blob/master/test/extended/testdata/router/router-http-echo-server.yaml). I would prefer to have `hello-openshift` available, but I am mentioning this method of creating an echo container if getting `hello-openshift` into the release image turns out to be problematic.

Reply:
I wish I'd caught this before (alas), but hello-openshift probably shouldn't have been chosen.  I would have expected the ingress operator image to contain an alternate command that exposed a specific HTTP server, and the operator image to start that variant.

I'd like to see us remove hello-openshift in a future release and switch to using a stub off the operator with the appropriate tooling.  Teams should use their operator images and operator commands as swiss army knives for this sort of internal testing.

Reply:
I see. @wking correct me if im wrong, but there was some other motivation besides this enhancement for adding hello-openshift to the relase image, correct? 
That being said, the ingress-operator could certainly switch over to an alternative command baked into the ingress-operators image. Just wondering if other components (cluster-samples?) would also be affected by the removal of hello-openshift from future release images.

Reply:
Trevor may have been referring to the docker problem - that was fixed in a different way (e2e tests verifying http should be using the agnhost image provided by kube), and hello-openshift was purged from there.

I don't know of any reason in product it should remain and I'd like to see it excised.  Operators doing testing should test using a known quantity (a subcommand they check), not a general purpose example image that can't be changed to exercise test behavior.

Reply:
hello-openshift can be summarily removed from the samples operator - if we haven't shipped it in 4.6 we *definitely* should not ship it in 4.7.  If we did ship it, we should mark it deprecated so no customer depends on it.

Reply:
The hello-openshift imagestream fixes the remaining use cases of "openshift/hello-openshift", both wrt pulling from Docker Hub and wrt multi-arch.  At least console and serverless currently still reference this in code and or tests.

Reply:
@yselkowitz would it be possible for the components you mentioned to switch over to an alternative image (via a sub-command, etc). for 4.8? Are we definitely removing hello-openshift in 4.8?

Reply:
as per @smarterclayton, hello-openshift will be removed in 4.8. I have updated this enhancement to explain that the ingress operator will pivot off of hello-openshift in future releases.